DSA Session/Assignment6/Graph Operations/src/graph/GraphImpl.java

```diff
@@ -0,0 +1,297 @@
+            for (int i = 0; i < vertices; i++) {
+                if (isVisited[i]) {
+                    listOfReachableNode.add(i + 1);
+                }
+
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            System.out.println(e.getMessage());
+        }
+        return listOfReachableNode;
+    }
+
+    @Override
+    public List<Edge> getMinimumSpanningTree() {
+        try {
+            boolean[] isVisited = new boolean[vertices];
+            getMSTUsingPrim(1, new ArrayList<Edge>(), isVisited);
+        } catch (Exception e) {
+            e.printStackTrace();
+            System.out.println(e.getMessage());
+        }
+        return listOfEdgesInMST;
+    }
+
+    /**
+     * Gets the minimum spanning tree using prim's algorithm.
+     * @param vertex : starting vertex.
+     * @param listOfTemporaryEdges : contains the spanning tree edges.
+     * @param isVisited : Keeps track of the visited nodes.
+     */
+    private void getMSTUsingPrim(int vertex, List<Edge> listOfTemporaryEdges,
+            boolean[] isVisited) {
+        try {
+            if (vertex <= 0 || vertex > vertices) {
+                throw new AssertionError("vertexNode doesn't exist in graph");
+            }
+
+            isVisited[vertex - 1] = true;
+            List<Edge> listOfAdjacentEdges = edgeList.get(vertex - 1);
+            // adjacent edges of vertex are added to the temporary list of edges
+
+            for (Edge edge : listOfAdjacentEdges) {
+
+                int destination = edge.getDestination();
+                if (!isVisited[destination - 1]) {
+                    listOfTemporaryEdges.add(edge);
+                }
+            }
+
+            Edge currentEdge = listOfTemporaryEdges.get(0);
+            int mininmumWeight = Integer.MAX_VALUE;
+            // finds edge with minimum weight and not exist in listOfEdgesInMST
+            for (Edge edge : listOfTemporaryEdges) {
+                int currentWeight = edge.getWeight();
+
+                boolean isMSTContainEdge = listOfEdgesInMST.contains(edge);
+
+                if (mininmumWeight > currentWeight && !isMSTContainEdge) {
+                    mininmumWeight = currentWeight;
+                    currentEdge = edge;
+                }
+            }
+
+            int destinationOfCurrent = currentEdge.getDestination();
+            if (!isVisited[destinationOfCurrent - 1]) {
+                listOfEdgesInMST.add(currentEdge);
+                listOfTemporaryEdges.remove(currentEdge);
+            }
+            for (int i = 0; i < isVisited.length; i++) {
+                // recursive call to function if any node is not visited.
+                if (!isVisited[i]) {
+                    getMSTUsingPrim(destinationOfCurrent, listOfTemporaryEdges,
+                            isVisited);
+                }
+
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            System.out.println(e.getMessage());
```

Comment:
I think you should use loggers here instead of sop.
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

    Logger logger = LoggerFactory.getLogger(this.getClass());

logger will provide you many methods to log your data. From next assignment you should use this.

Reply:
Okay Sir, I will read more about it and will use it.